viam-cartographer_internal_test.go

```diff
@@ -0,0 +1,348 @@
+// Package viamcartographer implements simultaneous localization and mapping.
+// This is an Experimental package.
+package viamcartographer
+
+import (
+	"bytes"
+	"context"
+	"testing"
+
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+	commonv1 "go.viam.com/api/common/v1"
+	v1 "go.viam.com/api/service/slam/v1"
+	"go.viam.com/rdk/services/slam"
+	"go.viam.com/rdk/spatialmath"
+	"go.viam.com/rdk/testutils/inject"
+	"go.viam.com/test"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+var (
+	chunkSizePointCloudMap     int
+	chunkSizeInternalState     int
+	inputComponentRef          string
+	expectedPointCloudMapBytes []byte
+	expectedInternalStateBytes []byte
+	inputQuat                  map[string]interface{}
+	inputPose                  commonv1.Pose
+)
+
+type pointCloudClientMock struct {
+	grpc.ClientStream
+	reader           *bytes.Reader
+	clientBufferSize int
+}
+
+func makePointCloudClientMock(b *[]byte, clientBufSize *int) *pointCloudClientMock {
+	return &pointCloudClientMock{reader: bytes.NewReader(*b), clientBufferSize: *clientBufSize}
+}
+
+func (m *pointCloudClientMock) Recv() (*v1.GetPointCloudMapResponse, error) {
+	clientBuffer := make([]byte, m.clientBufferSize)
+	n, err := m.reader.Read(clientBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetPointCloudMapResponse{
+		PointCloudPcdChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+type internalStateClientMock struct {
+	grpc.ClientStream
+	reader           *bytes.Reader
+	clientBufferSize int
+}
+
+func makeInternalStateClientMock(b *[]byte, clientBufSize *int) *internalStateClientMock {
+	return &internalStateClientMock{reader: bytes.NewReader(*b), clientBufferSize: *clientBufSize}
+}
+
+func (m *internalStateClientMock) Recv() (*v1.GetInternalStateResponse, error) {
+	clientBuffer := make([]byte, m.clientBufferSize)
+	n, err := m.reader.Read(clientBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetInternalStateResponse{
+		InternalStateChunk: clientBuffer[:n],
+	}
+	return resp, nil
+}
+
+func makeGetPositionResponse(
+	pose *commonv1.Pose,
+	quat *map[string]interface{},
+	componentReference *string,
+	quatKey string,
+) (*v1.GetPositionResponse, error) {
+	extra, err := structpb.NewStruct(map[string]interface{}{quatKey: *quat})
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &v1.GetPositionResponse{
+		Pose:               pose,
+		ComponentReference: *componentReference,
+		Extra:              extra,
+	}
+	return resp, nil
+}
+
+func makeQuaternionFromProto(quat map[string]interface{}) spatialmath.Orientation {
+	return &spatialmath.Quaternion{
+		Real: quat["real"].(float64),
+		Imag: quat["imag"].(float64),
+		Jmag: quat["jmag"].(float64),
+		Kmag: quat["kmag"].(float64),
+	}
+}
+
+func TestGetPositionEndpoint(t *testing.T) {
+	svc := &cartographerService{}
+	mockSLAMClient := &inject.SLAMServiceClient{}
+	svc.clientAlgo = mockSLAMClient
+
+	// Add successful GetPosition client
```

Comment:
[nit] I think you should wrap these tests in another `t.Run` instead of using comments. That should make it easier to debug. 
```suggestion
	t.Run("Successful GetPosition client", func(t *testing.T) {
```

Reply:
I'm personally not in favor of having multi-tier t.Runs for so few tests. Since the largest grouping of tests would 5x I dont think the additional organization is needed or looks as clean in the verbose logs. Below is an example of what the multi-tier might do to GetPosition tests:
```
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/success (0.00s)
        --- PASS: TestGetPositionEndpoint/success/origin_pose (0.00s)
        --- PASS: TestGetPositionEndpoint/success/non_origin_pose (0.00s)
        --- PASS: TestGetPositionEndpoint/success/empty_component_reference (0.00s)
    --- PASS: TestGetPositionEndpoint/failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/invalid_quat_map_keys_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/invalid_quat_map_values_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/no_quat_key_found_in_extra (0.00s)
        --- PASS: TestGetPositionEndpoint/failure/error_return (0.00s)
```
vs.
```
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/origin_pose_success (0.00s)
    --- PASS: TestGetPositionEndpoint/non_origin_pose_success (0.00s)
    --- PASS: TestGetPositionEndpoint/empty_component_reference_success (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_quat_map_keys_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_quat_map_values_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/no_quat_key_found_in_extra (0.00s)
    --- PASS: TestGetPositionEndpoint/error_return_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/undefined_GetPosition_function_in_injected_client (0.00s)
```

Reply:
@EshaMaharishi @nicksanford @kkufieta Let me know though if you like option 1 better!

Reply:
I don't feel strongly either way since there are few enough tests!

Reply:
New breakup:
```
=== RUN   TestGetPositionEndpoint
=== RUN   TestGetPositionEndpoint/successful_client
=== RUN   TestGetPositionEndpoint/successful_client/origin_pose_success
=== RUN   TestGetPositionEndpoint/successful_client/non_origin_pose_success
=== RUN   TestGetPositionEndpoint/successful_client/empty_component_reference_success
=== RUN   TestGetPositionEndpoint/successful_client/invalid_quat_map_keys_failure
=== RUN   TestGetPositionEndpoint/successful_client/invalid_quat_map_values_failure
=== RUN   TestGetPositionEndpoint/successful_client/empty_quat_map_failure
=== RUN   TestGetPositionEndpoint/invalid_client_no_quat_key_found_in_extra_failure
=== RUN   TestGetPositionEndpoint/invalid_client_error_return_failure
=== RUN   TestGetPositionEndpoint/nil_client_failure
--- PASS: TestGetPositionEndpoint (0.00s)
    --- PASS: TestGetPositionEndpoint/successful_client (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/origin_pose_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/non_origin_pose_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/empty_component_reference_success (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/invalid_quat_map_keys_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/invalid_quat_map_values_failure (0.00s)
        --- PASS: TestGetPositionEndpoint/successful_client/empty_quat_map_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_client_no_quat_key_found_in_extra_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetPositionEndpoint/nil_client_failure (0.00s)
=== RUN   TestGetPointCloudMapEndpoint
=== RUN   TestGetPointCloudMapEndpoint/successful_client
=== RUN   TestGetPointCloudMapEndpoint/successful_client/small_chunk_size_success
=== RUN   TestGetPointCloudMapEndpoint/successful_client/large_chunk_size_success
=== RUN   TestGetPointCloudMapEndpoint/successful_client/no_bytes_success
=== RUN   TestGetPointCloudMapEndpoint/invalid_client_error_return_failure
=== RUN   TestGetPointCloudMapEndpoint/nil_client_failure
--- PASS: TestGetPointCloudMapEndpoint (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/successful_client (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/small_chunk_size_success (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/large_chunk_size_success (0.00s)
        --- PASS: TestGetPointCloudMapEndpoint/successful_client/no_bytes_success (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetPointCloudMapEndpoint/nil_client_failure (0.00s)
=== RUN   TestGetInternalStateEndpoint
=== RUN   TestGetInternalStateEndpoint/successful_client
=== RUN   TestGetInternalStateEndpoint/successful_client/small_internal_state_chunk_size_success
=== RUN   TestGetInternalStateEndpoint/successful_client/large_internal_state_chunk_size_success
=== RUN   TestGetInternalStateEndpoint/successful_client/no_internal_state_bytes_success
=== RUN   TestGetInternalStateEndpoint/invalid_client_error_return_failure
=== RUN   TestGetInternalStateEndpoint/nil_client_failure
--- PASS: TestGetInternalStateEndpoint (0.00s)
    --- PASS: TestGetInternalStateEndpoint/successful_client (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/small_internal_state_chunk_size_success (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/large_internal_state_chunk_size_success (0.00s)
        --- PASS: TestGetInternalStateEndpoint/successful_client/no_internal_state_bytes_success (0.00s)
    --- PASS: TestGetInternalStateEndpoint/invalid_client_error_return_failure (0.00s)
    --- PASS: TestGetInternalStateEndpoint/nil_client_failure (0.00s)
PASS
```